File: python/treeList.py
### 1. Two Sum (Array & HashMap) - Python (Optimized)
def two_sum(nums, target):
    num_map = {}
    for i, num in enumerate(nums):
        complement = target - num
        if complement in num_map:
            return [num_map[complement], i]
        num_map[num] = i
    return None

### 2. Reverse a Linked List - Python (Optimized)
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverse_linked_list(head):
    prev, current = None, head
    while current:
        current.next, prev, current = prev, current, current.next
    return prev

### 3. Valid Parentheses - Python (Optimized)
def is_valid_parentheses(s):
    stack = []
    mapping = {')': '(', '}': '{', ']': '['}
    for char in s:
        if char in mapping:
            if not stack or stack.pop() != mapping[char]:
                return False
        else:
            stack.append(char)
    return not stack

### 4. Merge Intervals - Python (Optimized)
def merge_intervals(intervals):
    if not intervals:
        return []
    intervals.sort()
    merged = [intervals[0]]
    for start, end in intervals[1:]:
        if start <= merged[-1][1]:
            merged[-1][1] = max(merged[-1][1], end)
        else:
            merged.append([start, end])
    return merged

### 5. Binary Tree Level Order Traversal - Python (Optimized)
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def level_order_traversal(root):
    if not root:
        return []
    result, queue = [], deque([root])
    while queue:
        level = [node.val for node in queue]
        result.append(level)
        for _ in range(len(queue)):
            node = queue.popleft()
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
    return result

File: unusedGo/treeList.go
package main

import (
    "container/list"
    "fmt"
    "sort"
)

// 1. Two Sum - Optimized
func twoSum(nums []int, target int) []int {
    numMap := make(map[int]int)
    for i, num := range nums {
        if index, found := numMap[target-num]; found {
            return []int{index, i}
        }
        numMap[num] = i
    }
    return nil
}

// 2. Reverse a Linked List
 type ListNode struct {
    Val  int
    Next *ListNode
}

func reverseLinkedList(head *ListNode) *ListNode {
    var prev *ListNode
    current := head
    for current != nil {
        next := current.Next
        current.Next = prev
        prev = current
        current = next
    }
    return prev
}

// 3. Valid Parentheses
func isValidParentheses(s string) bool {
    stack := []rune{}
    mapping := map[rune]rune{')': '(', '}': '{', ']': '['}
    for _, char := range s {
        if open, exists := mapping[char]; exists {
            if len(stack) == 0 || stack[len(stack)-1] != open {
                return false
            }
            stack = stack[:len(stack)-1]
        } else {
            stack = append(stack, char)
        }
    }
    return len(stack) == 0
}

// 4. Merge Intervals
 type Interval struct {
    Start int
    End   int
}

func mergeIntervals(intervals []Interval) []Interval {
    if len(intervals) == 0 {
        return []Interval{}
    }
    sort.Slice(intervals, func(i, j int) bool {
        return intervals[i].Start < intervals[j].Start
    })
    merged := []Interval{intervals[0]}
    for _, interval := range intervals[1:] {
        last := &merged[len(merged)-1]
        if interval.Start <= last.End {
            last.End = max(last.End, interval.End)
        } else {
            merged = append(merged, interval)
        }
    }
    return merged
}

// 5. Binary Tree Level Order Traversal
 type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

func levelOrderTraversal(root *TreeNode) [][]int {
    if root == nil {
        return [][]int{}
    }
    result := [][]int{}
    queue := list.New()
    queue.PushBack(root)
    for queue.Len() > 0 {
        level := []int{}
        size := queue.Len()
        for i := 0; i < size; i++ {
            node := queue.Remove(queue.Front()).(*TreeNode)
            level = append(level, node.Val)
            if node.Left != nil {
                queue.PushBack(node.Left)
            }
            if node.Right != nil {
                queue.PushBack(node.Right)
            }
        }
        result = append(result, level)
    }
    return result
}
